### DISCUSSION TITLE
Show cause of "core commands are not available"

### DISCUSSION THREAD

**Discussion by amenk** [netz98/n98-magerun2#1512] [Ideas]
I would be cool to show the last error message of `bin/magento` in case the core commands are not available.

Especially in CI logs, this might allow quicker debugging.

**Comment by cmuench** [comment]
@amenk the latest n98-magerun2 version proxy all call to any core command. It's just calling bin/magento.
Error messages should be used.
Do you mean we should print a part of the error log? Maybe I do not fully understand your idea. Can you give me a hint?

**Reply by amenk** [reply]
I'm just talking about proxying - this should solve it

**Reply by amenk** [reply]
I did a test with magerun --unstable:

```
$ n98-magerun2 cache:flush
 Please verify if "bin/magento" is running.
Run with -vvv option to see the error output.
Only n98-magerun2 commands are available until the issue is fixed.

In Client.php line 508:
```

while

```
$ bin/magento 
Connection to Redis shop-example-com-redis-1:6379 failed after 2 failures.Last Error : (0) php_network_getaddresses: getaddrinfo for shop-example-com-redis-1 failed: Name or service not known#0 /home/amk/projects/shop.example.com/vendor/colinmollenhour/credis/Client.php(504): Credis_Client->connect()
#1 /home/amk/projects/shop.example.com/vendor/colinmollenhour/credis/Client.php(850): Credis_Client->connect()
#2 /home/amk/projects/shop.example.com/vendor/colinmollenhour/cache-backend-redis/Cm/Cache/Backend/Redis.php(485): Credis_Client->__call('hGet', Array)
#3 /home/amk/projects/shop.example.com/vendor/magento/zendframework1/library/Zend/Cache/Core.php(306): Cm_Cache_Backend_Redis->load('608_MAGE_VERSIO...', false)
#4 /home/amk/projects/shop.example.com/vendor/magento/framework/Cache/Frontend/Adapter/Zend.php(55): Zend_Cache_Core->load('608_MAGE_VERSIO...')
```


shows more detailled output.

n98-magerun2 -vvv also does not show that output from bin/magento. maybe STDERR is not piped through?

**Reply by cmuench** [reply]
Ok. Got it. I will check that.

**Reply by cmuench** [reply]
I was not able to reproduce it. I get the same output. The error output is respected.
My error also looks different as in your example (same cache:flush command).

That's the output on a current Mage-OS 1.0.6 with a not accessible redis cache server.

![image](https://github.com/user-attachments/assets/f119e400-037e-4343-ac18-a707bc3636a8)

**Reply by amenk** [reply]
okay, I seem to get the same output now. No clue what changed. in My first test I got the backtrace in bin/magento but not in n98-mager2. But what's more important:
If I use a command that is only defined in bin/magento I just get:


```
$ n98-magerun2 varnish:vcl:generate
Magento Core Commands cannot be loaded. Please verify if "bin/magento" is running.
Run with -vvv option to see the error output.
Only n98-magerun2 commands are available until the issue is fixed.

                                                                 
  There are no commands defined in the "varnish:vcl" namespace.  
```

So in this case I do not see any error message from bin/magento

```
$ bin/magento 
Connection to Redis standalone tcp://localhost:6379 failed after 2 failures.Last Error : (0) php_network_getaddresses: getaddrinfo for localhost failed: Name or service not known
```

So in a CI or so log this would be hard to debug.
We currently work around this by just directly calling `bin/magento` in such cases.

**Reply by cmuench** [reply]
That's an interesting finding.

**Reply by cmuench** [reply]
> That's an interesting finding.

We load the core commands by using "php bin/magento --format=json".
If this does not provide infos we do not know any core command. Could be that this is the cause of the difference.
I am not sure if we should then do a fallback with all arguments to "bin/magento" or supress the error message "There are no commands defined in the...".
I think supressing the misleading message is better. If bin/magento is broken then it's a broken core system.
